fv/wireguard_test.go

```diff
@@ -0,0 +1,455 @@
+					return out
+				}, "10s", "100ms").Should(BeEmpty())
+			}
+		})
+
+		It("the Wireguard rule-entry/route-table should exist", func() {
+			for i, felix := range felixes {
+				exists := wireguardRoutingRuleRouteEntryExists(felix, routingRuleRegExp, routeEntries[i])
+				Eventually(exists, "5s", "100ms").Should(BeTrue(), "wireguard rule missing")
+			}
+		})
+
+		It("v3 node resource annotations should contain public-keys", func() {
+			for _, felix := range felixes {
+				Eventually(func() string {
+					node, err := client.Nodes().Get(context.Background(), felix.Hostname, options.GetOptions{})
+					Expect(err).NotTo(HaveOccurred())
+					return node.Status.WireguardPublicKey
+				}, "5s", "100ms").ShouldNot(BeEmpty())
+			}
+		})
+
+		It("v3 node resource annotations should automatically heal", func() {
+			for _, felix := range felixes {
+				// Get the original public-key.
+				node, err := client.Nodes().Get(context.Background(), felix.Hostname, options.GetOptions{})
+				Expect(err).NotTo(HaveOccurred())
+				wgPubKeyOrig := node.Status.WireguardPublicKey
+
+				// overwrite public-key by fake but valid wireguard key.
+				node.Status.WireguardPublicKey = fakeWireguardPubKey
+				_, err = client.Nodes().Update(context.Background(), node, options.SetOptions{})
+				Expect(err).NotTo(HaveOccurred())
+
+				Eventually(func() string {
+					node, err := client.Nodes().Get(context.Background(), felix.Hostname, options.GetOptions{})
+					Expect(err).NotTo(HaveOccurred())
+					return node.Status.WireguardPublicKey
+				}, "5s", "100ms").Should(Equal(wgPubKeyOrig))
+			}
+		})
+
+		It("workload to workload traffic should be allowed and encrypted", func() {
+			for _, felix := range felixes {
+				// Start tcpdump on each "host".
+				tcpdump := felix.AttachTCPDump("eth0")
+				inTunnelPacketsPattern := fmt.Sprintf("IP %s\\.51820 > \\d+\\.\\d+\\.\\d+\\.\\d+\\.51820: UDP", felix.IP)
+				tcpdump.AddMatcher("numInTunnelPackets", regexp.MustCompile(inTunnelPacketsPattern))
+				outTunnelPacketsPattern := fmt.Sprintf("IP \\d+\\.\\d+\\.\\d+\\.\\d+\\.51820 > %s\\.51820: UDP", felix.IP)
+				tcpdump.AddMatcher("numOutTunnelPackets", regexp.MustCompile(outTunnelPacketsPattern))
+				tcpdump.Start()
+
+				tcpdumps = append(tcpdumps, tcpdump)
+			}
+
+			cc.ExpectSome(w[0], w[1])
+			cc.ExpectSome(w[1], w[0])
+			cc.CheckConnectivity()
+
+			By("verifying tunnelled packet count")
+			for i := range felixes {
+				Eventually(tcpdumpMatchCount(i, "numInTunnelPackets")).Should(BeNumerically(">", 0))
+				Eventually(tcpdumpMatchCount(i, "numOutTunnelPackets")).Should(BeNumerically(">", 0))
+			}
+
+			By("verifying peer tunnel transmit details are complimentary")
+			var sent, rcvd [nodeCount]string
+			for i, felix := range felixes {
+				out, _ := felix.ExecOutput("wg")
+				matches := xferRegExp.FindStringSubmatch(out)
+				Expect(matches).NotTo(BeNil())
+				rcvd[i] = matches[1]
+				sent[i] = matches[2]
+			}
+			Expect(rcvd[0]).To(Equal(sent[1]))
```

Comment:
this is a good idea.  Do we need to check that the wireguard interfaces are configured with the correct peers first though - otherwise isn't there a window where one peer could be configured, but not the other ... so you might get some asymmetric data.

Reply:
Probably not. Transmit data in `wg` output applies to each peer section. In a 2 node cluster, if the regex matches it is implied that that Wireguard tunnel is established with the only other pair in this case.

Reply:
@realgaurav - resolved?

Reply:
I guess this ok for now, but you should determine whether "probably not" is "actually not", otherwise I'm worried this might be a future source of flakes.  Should be easy enough to test - delete the peer config on the remote end and see if we still see the sent packets increasing, but no packets received.   If we do then we *may* want to 
tighten up the initial testing before we start sending packets, or work with delta packet counts.  Not sure.